src/graphql.rs

```diff
@@ -216,47 +216,100 @@ where
     E: Into<async_graphql::Error>,
 {
-    let (first, last) = validate_pagination_params(after.is_some(), before.is_some(), first, last)?;
+    let (first, last) = connection_size(after.is_some(), before.is_some(), first, last)?;
+
     async_graphql::connection::query(after, before, first, last, |after, before, first, last| {
         f(after, before, first, last)
     })
     .await
 }
 
-const DEFAULT_CONNECTION_SIZE: i32 = 100;
+async fn query_with_constraints<Node, ConnectionFields, Name, F, R, E>(
+    after: Option<String>,
+    before: Option<String>,
+    first: Option<i32>,
+    last: Option<i32>,
+    f: F,
+) -> Result<Connection<String, Node, ConnectionFields, EmptyFields, Name>>
+where
+    Node: OutputType,
+    ConnectionFields: ObjectType,
+    Name: ConnectionNameType,
+    F: FnOnce(Option<String>, Option<String>, Option<usize>, Option<usize>) -> R,
+    R: Future<Output = Result<Connection<String, Node, ConnectionFields, EmptyFields, Name>, E>>,
+    E: Into<async_graphql::Error>,
+{
+    extra_validate_pagination_params(
+        after.is_some(),
+        before.is_some(),
+        first.is_some(),
+        last.is_some(),
+    )?;
+    let (first, last) = connection_size(after.is_some(), before.is_some(), first, last)?;
 
-fn validate_pagination_params(
+    async_graphql::connection::query(after, before, first, last, |after, before, first, last| {
+        f(after, before, first, last)
+    })
+    .await
+}
+
+#[derive(Debug, thiserror::Error)]
+enum Error {
+    #[error("Invalid cursor")]
+    InvalidCursor,
+    #[error("The value of first and last must be within 1-100")]
+    InvalidLimitValue,
+    #[error("You must provide a `first` or `last` value to properly paginate a connection.")]
+    MissingPaginationBoundaries,
+    #[error("`after` and `before` should not be provided at the same time")]
+    InvalidPaginationArgumentsAfterBefore,
+    #[error("`after` and `last` should not be provided at the same time")]
+    InvalidPaginationArgumentsAfterLast,
+    #[error("`before` and `first` should not be provided at the same time")]
+    InvalidPaginationArgumentsBeforeFirst,
+    #[error("Missing validation")]
+    MissingValidation,
+}
+
+const DEFAULT_CONNECTION_SIZE: i32 = 10;
+const MAX_CONNECTION_SIZE: i32 = 100;
+
+fn connection_size(
     after: bool,
     before: bool,
-    mut first: Option<i32>,
-    mut last: Option<i32>,
-) -> Result<(Option<i32>, Option<i32>)> {
-    if let Some(first) = first {
-        if first < 0 {
-            return Err("The \"first\" parameter must be a non-negative number".into());
-        }
-    }
-    if let Some(last) = last {
-        if last < 0 {
-            return Err("The \"last\" parameter must be a non-negative number".into());
-        }
+    first: Option<i32>,
+    last: Option<i32>,
+) -> Result<(Option<i32>, Option<i32>), Error> {
+    match (after, before, first, last) {
+        (true, true, None, None) => Err(Error::MissingPaginationBoundaries),
```

Comment:
Is this error correct? It seems counter-intuitive to call it "**missing** pagination boundaries" when both `after` and `before` are `true`.

Reply:
I replaced it with `MissingPaginationSize`.

Reply:
The error message for `MissingPaginationSize` (“You must provide a first or last value to properly paginate a connection.”) implies that either first or last must always be provided. However, in the second and third `match` arms, neither `first` nor `last` is provided, and the function still accepts this by using a default size. This seems inconsistent with the error message. Either the error message should be updated to reflect the specific conditions that lead to the error (e.g., when both after and before are true), or the logic should enforce that one of first or last is always required, as the current error message suggests. Could you clarify this?

Reply:
I agree with your point. Based on your feedback, I adjusted how the connection size is handled:

- When users do not specify a connection size, it is assumed that they want to retrieve as many items as the server can provide. Therefore, `MAX_CONNECTION_SIZE` is used, and the previous `DEFAULT_CONNECTION_SIZE` has been removed.
- If both `after` and `before` are specified, users are likely not specifying a size because they want all items between the two cursors. To prevent performance degradation, `MAX_CONNECTION_SIZE` should be applied.
- As a result, `MissingPaginationSize` has been naturally removed.

What do you think?